### PR TITLE
fix(config): Use a safe regex library rather than sandboxing builtins.

### PR DESCRIPTION
**Pull request by rfk**
I may have gotten a little..uh..nerd-sniped this afternoon, but I took at look at alternatives for https://github.com/mozilla/fxa/issues/4403 and I thought to myself: *"if only we could use the regex crate from rust, which is specifically designed to work with untrusted inputs!"*.

Less than an hour of playing around with `wasm-pack` later, I seem to have a minimal PoC that works as intended:

  https://github.com/rfk/sandboxed-regexp

I don't think we should land this as-is, but with a bit of polish, I think it could actually be a safer and faster approach than what we're currently doing with sandboxed execution.

What do you think, worth pursuing further, or just-too-wacky-to-use-in-practice?

---

Because:

* We want to safely work with regexes from potentially-untrusted sources.
* The native JS RegExp class is vulnerable to DoS under such circumstances.
* Our existing sandboxing solution uses a library with known escape issues.
* It's fun to play with the rust+wasm toolchain.

This commit:

* Replaces our bespoke evaluate-regexp-in-a-sandbox solution with
  a wasm-compiled version of the Rust `regex` crate, which is
  specifically designed for working with untrusted regexes.

### CODE REVIEW

File: packages/fxa-auth-server/test/local/senders/select_email_services.js [hunk 1]
```diff
@@ -719,173 +653,6 @@ describe('selectEmailServices:', () => {
...
-      },
-    ]);
-
-    assert.equal(safeRegex.callCount, 1);
-    const args = safeRegex.args[0];
-    assert.equal(args.length, 1);
-    assert.equal(args[0], '^((((.*)*)*)*)*@example\\.com$');
-  });
-});
-
```
Comment (rfk) [inline packages/fxa-auth-server/test/local/senders/select_email_services.js]: Look at all the tests you don't need when your regex library doesn't have any such thing as an "unsafe regex"!

**Comment by dannycoates** [comment]
Pretty effing sweet @rfk 🤘

**Comment by rfk** [comment]
> Maybe we should move it to the Mozilla org, so you aren't permanently on the hook for solo maintenance.

Yep, very happy to do so, just didn't want to make speculative `mozilla` repo too early in the game 👍 

I poked at this a little more today. I was worried that the overhead of making a wasm module and its more expensive regexp compilation procedure would be too much, so I made a tiny bad comparison with the old "spawn a process in a sandbox" approach using a very simple regex. It seems consistently 60 to 100 times on my machines, which seems promising :-)

@dannycoates @6a68 if you're interested, please try pulling latest master from https://github.com/rfk/sandboxed-regex and running `make bench`. I'd be curious to see what the numbers look like on someone else's machine, just for a gutcheck.

**Comment by 6a68** [comment]
Typo: the repo is https://github.com/rfk/sandboxed-regexp. Took me a minute to find it.

Note that I had to manually `npm install sandbox` to get `make bench` to run. I got about 70x faster results from `make bench` across 3 runs.

I could see Glean possibly using the rust-wasm approach for maintaining a single codebase that runs in web content, so this might be a good first step to explore how wasm-rust works out in production.

**Comment by rfk** [comment]
To my surprise, I think this is actually ready for review.

First, you'll want to take a peek at the [sandboxed-regexp](https://github.com/rfk/sandboxed-regexp) repo. There's not a lot of code in there but it's worth reviewing to re-assure yourself that what it's doing makes sense, and considering the [list of caveats](https://github.com/rfk/sandboxed-regexp#being-careful-with-it). If it seems like a good path forward, we can move it under the mozilla org and publish a new version with more co-maintainers.

With that available from npm, the diff here is mostly just a cleanup! You will definitely want to view it with whitespace changes hidden, otherwise git makes an absolute dog's breakfast of the diff.

File: packages/fxa-auth-server/lib/senders/select_email_services.js [hunk 2]
```diff
@@ -94,115 +163,60 @@ module.exports = (log, config, mailer, emailService) => {
...
+        liveConfig = lastSeenLiveConfigParsed;
+      } else {
+        try {
+          lastSeenLiveConfig = liveConfig;
+          liveConfig = LiveConfig.parse(liveConfig);
+          lastSeenLiveConfigParsed = liveConfig;
+        } catch (err) {
+          lastSeenLiveConfig = lastSeenLiveConfigParsed = null;
+          log.error('emailConfig.parse.error', { err: err.message });
+        }
```
Comment (rfk) [inline packages/fxa-auth-server/lib/senders/select_email_services.js:176]: The new sandboxing approach seems uniformly faster than the old one in our tests, so there's technically no need for this extra caching logic here. But it just seems like an easy win for a value that doesn't change very often and can be safely cached in memory.

(Also, creating many `SandboxedRegExp` instances at once will allocate lots of memory, which the garbage-collector will probably stay on top of, but still...we can easily save it some work here).

File: packages/fxa-auth-server/lib/senders/select_email_services.js [hunk 3]
```diff
@@ -94,115 +163,60 @@ module.exports = (log, config, mailer, emailService) => {
...
       }
     }
 
-    const services = await emailAddresses.reduce(
-      async (promise, emailAddress) => {
```
Comment (rfk) [inline packages/fxa-auth-server/lib/senders/select_email_services.js:105]: Not sad to see the end of this very clever, but terrifying nested-promise-reducer-loop. Being able to test the regexes synchronously is a nice side benefit of this change.

**Comment by bbangert** [comment]
@rfk Looks like there's some test failures here.

File: packages/fxa-auth-server/lib/senders/select_email_services.js [hunk 4]
```diff
@@ -17,6 +16,98 @@ const SERVICES = {
...
+// Where a percentage and a regex are both present, an email address must
+// satisfy both criteria to count as a match. Where an email address matches
+// sendgrid and ses, sendgrid wins. Where an email address matches socketlabs
+// and ses, socketlabs wins. Where an email address matches sendgrid and
+// socketlabs, sendgrid wins.
+//
+// Regular expressions are executed using the `SandboxedRegExp` class in order
+// to protect againt DoS if a malicious regex happened to make its way into the
+// live config; as a result the regex syntax may differ from that of standard
+// javascript, but you shouldn't be putting complex regex in here anyway.
```
Comment (dannycoates) [inline packages/fxa-auth-server/lib/senders/select_email_services.js:45]: I'd like a link to https://docs.rs/regex/1.3.6/regex/#syntax or even https://github.com/rfk/sandboxed-regexp#being-careful-with-it here

File: packages/fxa-auth-server/lib/senders/select_email_services.js [hunk 5]
```diff
@@ -17,6 +16,98 @@ const SERVICES = {
...
+
+class LiveConfigForSender {
+  constructor(props) {
+    const { percentage, regex } = props;
+    if (percentage && percentage >= 0 && percentage < 100) {
+      this.percentage = percentage;
+    } else {
+      this.percentage = null;
```
Comment (dannycoates) [inline packages/fxa-auth-server/lib/senders/select_email_services.js:53]: nit: from a types perspective it'd be nice if `this.percentage` was just `number` not `number | null` (in a distant typescript future)

Comment (rfk) [reply packages/fxa-auth-server/lib/senders/select_email_services.js:53]: makes sense, I'll default it to `0` rather than `null`.

**Comment by rfk** [comment]
> If it seems like a good path forward, we can move it under the mozilla org and publish a new
> version with more co-maintainers.

I'm going to go ahead and take a look at transferring this repo.

**Comment by rfk** [comment]
I'd also like to find a way to test this new implementation against our current live config, just in case. @jrgm are you able to suggest an easy way to do that, which doesn't involve sending me the current config values (which I assume contain PII)?

I could for example prep a stand-alone script that uses the new sandboxing code, for someone from ops to try out manually.

**Comment by jrgm** [comment]
We actually haven't used this for any particularly tortuous regex patterns in email routing so you can just pick your own test patterns (a couple of emails, now retired from use).

File: packages/fxa-auth-server/lib/senders/select_email_services.js [hunk 6]
```diff
@@ -17,6 +16,101 @@ const SERVICES = {
...
+//     regex: "^.+@example\.com$"
+//   },
+//   socketlabs: {
+//     percentage: 100,
+//     regex: "^.+@example\.org$"
+//   },
+//   ses: {
+//     percentage: 10,
+//     regex: ".*",
+//   }
```
[the remark this answers is not in the document]
Comment (rfk) [reply packages/fxa-auth-server/lib/senders/select_email_services.js:33]: It does have a fair few tests already to cover the various cases, and I'm happy to not have had to change any of them as part of this PR 😄

**Comment by rfk** [comment]
Thanks for the feedback all, I've rebased this and will go ahead and merge if it's green.

It sounds the risk of this going sideways in prod will be fairly low, but a couple of things we should watch out for:

* If the new `SandboxedRegexp` class starts throwing errors, I don't see any "catch errors and default to a fallback" logic, so they'll probably show up as 500 errors on email sending routes, which I'm sure we'll notice quickly i prod.
* If the new logic turns out not to process things correctly in production, it'll show up as a change in the relative volume of emails being sent by different backends. I'm not sure how reliably or quickly we'd detect such an issue in production, but it'll be worth keeping an eye on when this goes live.

**Pull request merged**